File: FunTools.UnitTests/NonBlockingDownloadAnyOfTwoSitesWithWebClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using NUnit.Framework;

namespace FunTools.UnitTests
{
	[TestFixture]
	public class NonBlockingDownloadAnyOfTwoSitesWithWebClient
	{
		[Test]
		public void I_can_make_NonBlockingDownloadAnyOfTwoSitesWithWebClientTest()
		{
            var urls = new[] { "http://www.codeproject.com/", "http://smellegantcode.wordpress.com/" };

			var success = urls.Select(DownloadAsync)
				.AwaitSome((x, _) => Some.Of(x.Success))
				.WaitSuccess();

            Assert.That(success, Is.StringContaining("codeproject").Or.StringContaining("smell"));
		}

		[Test]
		public void If_both_sites_downloads_are_successfull_Then_result_should_contain_first_awaited_download()
		{
            var urls = new[] { "http://www.codeproject.com/", "http://smellegantcode.wordpress.com/" };
			var errors = new List<Exception>();

			var result = urls.Select(DownloadAsync)
				.AwaitSome(x => x.OnError(errors.Add).ConvertTo(Some.Of, None.Of<string>()))
				.WaitSuccess();

            Assert.That(result, Is.StringContaining("codeproject").Or.StringContaining("smell"));
		}

		[Test]
		public void If_first_site_download_is_failed_Then_result_should_contain_second_successful_download()
		{
			var errors = new List<Exception>();

			var result = Await.Many(
				(x, _) => x.OnError(errors.Add).ConvertTo(Some.Of, None.Of<string>()),
				null,
				DownloadAsync("http://דûדû.com"),
                DownloadAsync("http://www.codeproject.com/"))
				.WaitSuccess();

            Assert.That(result, Is.StringContaining("codeproject"));
		}

		[Test]
		public void If_both_sites_downloads_are_failed_Then_result_should_be_none_and_errors_list_should_contain_two_items()
		{
			var errors = new List<Exception>();

			var result = Await.Many(
				(x, _) => x.OnError(errors.Add).ConvertTo(Some.Of, ex => None.Of<string>()),
				null,
				DownloadAsync("http://דûדû.com"),
				DownloadAsync("http://ץוץו.com"))
				.WaitSuccess();

			Assert.That(result, Is.Null);
            Assert.That(errors.Count, Is.EqualTo(2));
		}

		#region Implementation

		private static Await<string> DownloadAsync(string uri)
		{
			var url = new Uri(uri);
			return complete =>
			{
				var webClient = new WebClient();

				var awaitDownload = Await.Event<DownloadDataCompletedEventArgs, DownloadDataCompletedEventHandler, string>(
					e => Result.Of(Encoding.ASCII.GetString(e.Result), e.Error).Success,
					h => webClient.DownloadDataCompleted += h,
					h => webClient.DownloadDataCompleted -= h,
					a => a.Invoke);

				var download = awaitDownload(complete);

				webClient.DownloadDataAsync(url);

				return download;
			};
		}

		#endregion
	}
}

File: FunTools.UnitTests.FS/NonBlockingDownloadAnyOfTwoSitesWithWebClient.fs
module FunTools.Tests.``Download two sites in parallel and handle possible download errors``

open FunTools
open NUnit.Framework
open FsUnit
open System
open System.Net

let downloadAsync uri =
    let url = new Uri(uri)
    Await(fun complete ->
        let webClient = new Net.WebClient()
        let download = 
            Await.Event<DownloadDataCompletedEventArgs, DownloadDataCompletedEventHandler, string>(
                (fun e ->
                    if e.Error <> null then e.Error.ReThrow()
                    Some.Of(Text.Encoding.ASCII.GetString e.Result)),
                (fun h -> webClient.DownloadDataCompleted.AddHandler(h)),
                (fun h -> webClient.DownloadDataCompleted.RemoveHandler(h)),
                (fun a -> DownloadDataCompletedEventHandler(fun o e -> a.Invoke(o, e))))
                .Invoke(complete)
        
        webClient.DownloadDataAsync(url)
        download)

[<Test>]
let ``When one site download fails but another succeeds Then result should contain another one`` () =
    let errors = ref []
    let result = 
        Await.Many(
            (fun (x : Result<_>) _ ->
                if x.IsSuccess then Some.Of(x.Success) 
                else
                    errors := x.Failure :: !errors
                    None.Of<string>()),
            null,
            [|"http://гыгы.com"; "http://www.infoq.com"|] |> Array.map downloadAsync
            ).WaitSuccess()
    
    result |> should contain "infoq"
    !errors |> should haveLength 1